core/src/main/java/org/apache/iceberg/rest/RestCatalog.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.rest.http.ErrorHandlers;
+import org.apache.iceberg.rest.http.HttpClient;
+import org.apache.iceberg.rest.http.RequestResponseSerializers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RestCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces, Configurable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RestCatalog.class);
+
+  private String catalogName;
+  private Map<String, String> properties;
+  private ObjectMapper mapper;
+  private String baseUrl;
+  private Configuration hadoopConf;
+  private FileIO fileIO;
+
+  // TODO - Refactor to use interface, which doesn't currently have POST etc embedded into it.
+  private HttpClient restClient;
+
+  @Override
+  public void initialize(String name, Map<String, String> props) {
+    super.initialize(name, props);
+    this.catalogName = name;
+    this.properties = props;
+
+    // TODO - Possibly authenticate with the server initially and then have the server return some of this information
+    Preconditions.checkNotNull(
+        properties.getOrDefault("baseUrl", null),
+        "Cannot initialize the RestCatalog as the baseUrl is a required parameter.");
+
+    this.baseUrl = properties.get("baseUrl");
+    this.fileIO = initializeFileIO(properties);
+
+    this.mapper = new ObjectMapper();
+    RequestResponseSerializers.registerAll(mapper);
+
+    // TODO - We can possibly handle multiple warehouses via one RestCatalog to reuse the connection pool
+    //        and for cross database calls if users need to authenticate with each.
+    restClient = HttpClient.builder()
+        .baseUrl(String.format("%s/warehouse/%s", baseUrl, catalogName))
+        .mapper(mapper)
+        .defaultErrorHandler(ErrorHandlers.tableErrorHandler())
+        .build();
+  }
+
+  @Override
+  public String name() {
+    return catalogName;
+  }
+
+  @Override
+  public Table createTable(
+      TableIdentifier identifier, Schema schema, PartitionSpec spec, String location,
+      Map<String, String> props) {
+    throw new UnsupportedOperationException("Not implemented: createTable");
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    // Users might have not authenticated, possibly need to reauthenticate, or get authentication info per request.
+    //   Though usually, auth tokens are persistent for at least some period of time.
+    return newTableOps(tableIdentifier, null);
+  }
+
+  protected RestTableOperations newTableOps(TableIdentifier tableIdentifier, String authToken) {
+    throw new UnsupportedOperationException("Not implemented: newTableOps");
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
```

Comment:
I'm not sure that we want to implement `BaseMetastoreCatalog`, since it has conventions that we don't need, like this method. This method is a way for the catalog to determine the default location for a table. But shouldn't this be done on the server side for a REST-based catalog? That makes the most sense to me because we want to keep the client as simple as possible and do all of the customization in the server that is easier to change. Deployed client Jars are hard to update!

Reply:
That makes sense to me. For the REST catalog, especially as a spec, the server should have a bit more control over what's done. I had a version that didn't use `BaseMetastoreCatalog` (and still do), but decided to forgo it initially given the deviation from the norm.

I'll work on switching it back as I'm fleshing out ideas for the spec.